File: Student.java
import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.SQLException;
import java.util.ArrayList;
import java.util.List;

public class Student {
    private String name;
    private String surname;
    private int age;
    private String specialNeeds;
    private List<Activity> selectedActivities; // Changed to store Activity objects
    private double totalExpenses;
    private double feeShare;
    private double balance; // Positive: to receive, Negative: to reimburse

    public Student(String name, String surname, int age, String specialNeeds) {
        this.name = name;
        this.surname = surname;
        this.age = age;
        this.specialNeeds = specialNeeds;
        this.selectedActivities = new ArrayList<>();
        this.totalExpenses = 0.0;
        this.feeShare = 0.0;
        this.balance = 0.0;
    }

    public void addActivity(Activity activity) { // Updated to accept Activity objects
        try {
            if (activity.isFull()) {
                System.out.println("Activity '" + activity.getName() + "' is already full.");
                return;
            }
        } catch (Exception e) {
            // TODO Auto-generated catch block
            e.printStackTrace();
        }
        selectedActivities.add(activity);
    }

    public List<Activity> getSelectedActivities() { // Updated return type
        return selectedActivities;
    }

    public void addExpense(double amount) {
        totalExpenses += amount;
    }

    public double getTotalExpenses() {
        return totalExpenses;
    }

    public void setFeeShare(double feeShare) {
        this.feeShare = feeShare;
    }

    public double getFeeShare() {
        return feeShare;
    }

    public void setBalance(double balance) {
        this.balance = balance;
    }

    public double getBalance() {
        return balance;
    }

    public Student(int id) {
    this.id = id;
}

    public void saveToDatabase() {
        try (Connection connection = DbConnection.connect()) {
            String sql = "INSERT INTO students (name, surname, age, special_needs, total_expenses, fee_share, balance) VALUES (?, ?, ?, ?, ?, ?, ?)";
            PreparedStatement statement = connection.prepareStatement(sql);
            statement.setString(1, name);
            statement.setString(2, surname);
            statement.setInt(3, age);
            statement.setString(4, specialNeeds);
            statement.setDouble(5, totalExpenses);
            statement.setDouble(6, feeShare);
            statement.setDouble(7, balance);
            statement.executeUpdate();
            System.out.println("Student saved to database.");
        } catch (SQLException e) {
            System.err.println("Error saving student to database: " + e.getMessage());
        }
    }

    @Override
    public String toString() {
        return "Student [name=" + name + ", surname=" + surname + ", age=" + age + ", specialNeeds=" + specialNeeds
                + ", selectedActivities=" + selectedActivities + ", totalExpenses=" + totalExpenses + ", feeShare="
                + feeShare + ", balance=" + balance + "]";
    }
}

File: student.py
from PythonExpenseApp.db_connection import DbConnection  # Importa la classe per la connessione al database
import mysql.connector  # Importa il modulo per la connessione a MySQL

class Student:
    def __init__(self, name, surname, age, special_needs):
        self.name = name  # Nome dello studente
        self.surname = surname  # Cognome dello studente
        self.age = age  # Età dello studente
        self.special_needs = special_needs  # Esigenze particolari (es. alimentari)
        self.selected_activities = []  # Lista delle attività selezionate dallo studente
        self.total_expenses = 0.0  # Totale delle spese sostenute dallo studente
        self.fee_share = 0.0  # Quota di spesa da dividere
        self.balance = 0.0  # Saldo finale (positivo o negativo)

    def add_activity(self, activity):
        try:
            if activity.is_full():  # Controlla se l'attività è già piena
                print(f"Activity '{activity.name}' is already full.")
                return
        except Exception as e:
            print(f"Error checking activity capacity: {e}")
        self.selected_activities.append(activity)  # Aggiunge l'attività alla lista

    def get_selected_activities(self):
        return self.selected_activities  # Restituisce la lista delle attività selezionate

    def add_expense(self, amount):
        self.total_expenses += amount  # Aggiunge una spesa al totale

    def get_total_expenses(self):
        return self.total_expenses  # Restituisce il totale delle spese

    def set_fee_share(self, fee_share):
        self.fee_share = fee_share  # Imposta la quota di spesa

    def get_fee_share(self):
        return self.fee_share  # Restituisce la quota di spesa

    def set_balance(self, balance):
        self.balance = balance  # Imposta il saldo

    def get_balance(self):
        return self.balance  # Restituisce il saldo

    def __init__(self, id):
        self.id = id  # Costruttore alternativo che imposta solo l'id (usato per caricamento da DB)

    def save_to_database(self): # Salva lo studente nel database MySQL
        connection = DbConnection.connect()  # Ottiene la connessione al database
        if connection: 
            try:
                sql = """INSERT INTO students
                         (name, surname, age, special_needs, total_expenses, fee_share, balance)
                         VALUES (%s, %s, %s, %s, %s, %s, %s)"""  # Query per inserire uno studente
                statement = connection.cursor() # Crea un cursore per eseguire la query
                statement.execute(sql, (self.name, self.surname, self.age, self.special_needs,
                                        self.total_expenses, self.fee_share, self.balance))
                connection.commit() # Salva le modifiche nel database
                print("Student saved to database.")  # Messaggio di conferma
            except mysql.connector.Error as e:
                print(f"Error saving student to database: {e}")  # Gestione errori

    def update_in_database(self): # Aggiorna i dati dello studente nel database MySQL
        connection = DbConnection.connect()  # Ottiene la connessione al database
        if connection:
            try:
                sql = """UPDATE students
                         SET total_expenses = %s, fee_share = %s, balance = %s
                         WHERE name = %s AND surname = %s"""  # Query per aggiornare i dati dello studente
                statement = connection.cursor()
                statement.execute(sql, (self.total_expenses, self.fee_share, self.balance,
                                        self.name, self.surname))
                connection.commit()
                print("Student updated in database.")  # Messaggio di conferma
            except mysql.connector.Error as e:
                print(f"Error updating student in database: {e}")  # Gestione errori

    @staticmethod # Metodo statico per ottenere tutti gli studenti dal database
    def get_all_students(): #   Recupera tutti gli studenti dal database MySQL
        students = []  # Lista di studenti da restituire
        connection = DbConnection.connect()  # Ottiene la connessione al database
        if connection: # Se la connessione è riuscita
            try:
                sql = """SELECT name, surname, age, special_needs, total_expenses,
                                fee_share, balance FROM students"""  # Query per recuperare tutti gli studenti
                statement = connection.cursor() # Crea un cursore per eseguire la query
                statement.execute(sql) # Esegue la query
                for (name, surname, age, special_needs, total_expenses, fee_share, balance) in statement: # Itera sui risultati
                    stu = Student(name, surname, age, special_needs)  # Crea un oggetto Student
                    stu.add_expense(total_expenses)  # Imposta il totale spese
                    stu.set_fee_share(fee_share)  # Imposta la quota spese
                    stu.set_balance(balance)  # Imposta il saldo
                    students.append(stu)  # Aggiunge lo studente alla lista
            except mysql.connector.Error as e: # Gestione errori MySQL
                print(f"Error retrieving students from database: {e}")  # Gestione errori
        return students  # Restituisce la lista di studenti

    def __str__(self):
        # Rappresentazione testuale dell'oggetto Student
        return (f"Student [name={self.name}, surname={self.surname}, age={self.age}, "
                f"specialNeeds={self.special_needs}, selectedActivities={self.selected_activities}, "
                f"totalExpenses={self.total_expenses}, feeShare={self.fee_share}, "
                f"balance={self.balance}]")